src/ngraph/node_output.hpp

```diff
@@ -0,0 +1,91 @@
+//*****************************************************************************
+// Copyright 2017-2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#pragma once
+
+#include "ngraph/descriptor/input.hpp"
+#include "ngraph/node.hpp"
+
+namespace ngraph
+{
+    /// \brief A handle for one of a node's outputs.
+    class NodeOutput
```

Comment:
I would love to see both NodeInput and NodeOutput derived some the same common abstract base class. There is so much in common or that should be in common that it really would make things nice I think.

Reply:
This is an interesting idea (maybe something like `NodePort` with `is_input` and `is_output` virtuals? though the virtual means every one of these objects now has its own vtable)---let me think about it and get back to you.

Reply:
I am a bit against it unless the abstract class is some internal class; the implementations are similar, but the API/semantics are separate.